source/windows/secure_channel_tls_handler.c

```diff
@@ -939,73 +935,105 @@ static int s_do_client_side_negotiation_step_2(struct aws_channel_handler *handl
+            .BufferType = SECBUFFER_DATA,
+        };
 
-    SecBufferDesc buffer_desc = {
-        .ulVersion = SECBUFFER_VERSION,
-        .cBuffers = 4,
-        .pBuffers = input_buffers,
-    };
+        SecBufferDesc buffer_desc = {
+            .ulVersion = SECBUFFER_VERSION,
+            .cBuffers = 4,
+            .pBuffers = input_buffers,
+        };
 
-    SECURITY_STATUS status = DecryptMessage(&sc_handler->sec_handle, &buffer_desc, 0, NULL);
+        sc_handler->read_extra = 0;
 
-    if (status == SEC_E_OK) {
-        /* if SECBUFFER_DATA is the buffer type of the second buffer, we have decrypted data to process.
-           If SECBUFFER_DATA is the type for the fourth buffer we need to keep track of it so we can shift
-           everything before doing another decrypt operation.
-           As far as I can tell, we don't care what's in the third buffer for TLS usage.*/
-        if (input_buffers[1].BufferType == SECBUFFER_DATA) {
-            size_t decrypted_length = input_buffers[1].cbBuffer;
-            AWS_LOGF_TRACE(
-                AWS_LS_IO_TLS,
-                "id=%p: Decrypted message with length %llu.",
-                (void *)handler,
-                (unsigned long long)decrypted_length);
+        SECURITY_STATUS status = DecryptMessage(&sc_handler->sec_handle, &buffer_desc, 0, NULL);
 
-            if (input_buffers[3].BufferType == SECBUFFER_EXTRA) {
-                sc_handler->read_extra = input_buffers[3].cbBuffer;
+        if (status == SEC_E_OK) {
+            /* if SECBUFFER_DATA is the buffer type of the second buffer, we have decrypted data to process.
+               If SECBUFFER_DATA is the type for the fourth buffer we need to keep track of it so we can shift
+               everything before doing another decrypt operation.
+               As far as I can tell, we don't care what's in the third buffer for TLS usage.*/
+            if (input_buffers[1].BufferType == SECBUFFER_DATA) {
+                size_t decrypted_length = input_buffers[1].cbBuffer;
                 AWS_LOGF_TRACE(
                     AWS_LS_IO_TLS,
-                    "id=%p: Extra (incomplete) message received with length %llu.",
+                    "id=%p: Decrypted message with length %llu.",
                     (void *)handler,
-                    (unsigned long long)sc_handler->read_extra);
+                    (unsigned long long)decrypted_length);
+
+                assert(
+                    decrypted_length <=
+                    sc_handler->buffered_read_out_data_buf.capacity - sc_handler->buffered_read_out_data_buf.len);
+                /* take what was decrypted and append it to the output buffers.*/
+                memcpy(
+                    sc_handler->buffered_read_out_data_buf.buffer + sc_handler->buffered_read_out_data_buf.len,
+                    (uint8_t *)input_buffers[1].pvBuffer,
+                    decrypted_length);

```

Comment:
just fyi, could have done this:
```
bool written = aws_byte_buf_append(&sc_handler->buffered_read_out_data_buf, input_buffers[1].pvBuffer, decrypted_length);
assert(written);
```
instead of this:
```
assert(decrypted_length <= sc_handler->buffered_read_out_data_buf.capacity - sc_handler->buffered_read_out_data_buf.len);
memcpy(sc_handler->buffered_read_out_data_buf.buffer + sc_handler->buffered_read_out_data_buf.len,  (uint8_t *)input_buffers[1].pvBuffer, decrypted_length);
sc_handler->buffered_read_out_data_buf.len += decrypted_length;
```